src/device_queue.cc

```diff
@@ -48,26 +48,42 @@ namespace detail {
 			} else {
 				const auto host_cfg = cfg.get_host_config();
 
-				// Try to find a platform that can provide a unique device for each node.
-				bool found = false;
-				const auto platforms = cl::sycl::platform::get_platforms();
-				for(size_t i = 0; i < platforms.size(); ++i) {
-					auto&& platform = platforms[i];
-					const auto devices = platform.get_devices(cl::sycl::info::device_type::gpu);
-					if(devices.size() >= host_cfg.node_count) {
-						how_selected = fmt::format("automatically selected platform {}, device {}", i, host_cfg.local_rank);
-						device = devices[host_cfg.local_rank];
-						found = true;
-						break;
+				const auto try_find_device_per_node = [&host_cfg, &device, &how_selected](cl::sycl::info::device_type type) {
+					// Try to find a platform that can provide a unique device for each node.
+					const auto platforms = cl::sycl::platform::get_platforms();
+					for(size_t i = 0; i < platforms.size(); ++i) {
+						auto&& platform = platforms[i];
+						const auto devices = platform.get_devices(type);
+						if(devices.size() >= host_cfg.node_count) {
+							how_selected = fmt::format("automatically selected platform {}, device {}", i, host_cfg.local_rank);
+							device = devices[host_cfg.local_rank];
+							return true;
+						}
 					}
-				}
+					return false;
+				};
 
-				if(!found) {
-					queue_logger.warn("No suitable platform found that can provide {} devices, and CELERITY_DEVICES not set", host_cfg.node_count);
-					// Just use the first device available
-					const auto devices = cl::sycl::device::get_devices(cl::sycl::info::device_type::gpu);
-					if(devices.empty()) { throw std::runtime_error("Automatic device selection failed: No GPU device available"); }
-					device = devices[0];
+				const auto try_find_one_device = [&device](cl::sycl::info::device_type type) {
```

Comment:
This path does not set `how_selected`.

Reply:
`how_selected` is initialized to `"automatically selected"` on line 29 (yes not ideal, we'll address it during refactoring ;-)).